library/Fission/User/Modifier.hs

```diff
@@ -36,13 +42,65 @@ updatePublicKeyDB ::
   => UserId
   -> Key.Public
   -> UTCTime
-  -> Transaction m ()
-updatePublicKeyDB userID pk now =
+  -> Transaction m (Either Errors Key.Public)
+updatePublicKeyDB userID pk now = do
   update userID
     [ UserPublicKey  =. Just pk
     , UserModifiedAt =. now
     ]
-  
+
+  return $ Right pk
+
+addExchangeKeyDB ::
+     MonadIO m
+  => UserId
+  -> RSA.PublicKey
+  -> UTCTime
+  -> Transaction m (Either Errors [RSA.PublicKey])
+addExchangeKeyDB userID key now =
+  Persist.get userID >>= \case
+    Nothing -> 
+      return . openLeft $ NotFound @User
+
+    Just user -> do
+      let keys = userExchangeKeys user
+      if List.elem key keys
+        then 
+          return $ Right keys
+
+        else do
+          let updated = keys ++ [key]
+          update userID
+            [ UserExchangeKeys =. updated
+            , UserModifiedAt   =. now
+            ]
+          return $ Right updated
+
+removeExchangeKeyDB ::
+    MonadIO m
+  => UserId
+  -> RSA.PublicKey
+  -> UTCTime
+  -> Transaction m (Either Errors [RSA.PublicKey])
+removeExchangeKeyDB userID key now =
+  Persist.get userID >>= \case
+    Nothing -> 
+      return . openLeft $ NotFound @User
+
+    Just user -> do
+      let keys = userExchangeKeys user
+      if List.elem key keys
+        then do
+          let updated = List.delete key keys
+          update userID
+            [ UserExchangeKeys =. updated
+            , UserModifiedAt   =. now
+            ]
+          return $ Right updated
+
+        else 
+          return $ Right keys
```

Comment:
No need to change, but a difference from say JS is that the following is totally valid and 

```haskell
let
  keys = userExchange user
  updated = key `List.delete` keys
  
if List.elem key keys
  then do
   -- ...
```

We never actually call `updated` until it’s needed, and since `let`s are totally pure they can’t do anything to the `keys` binding, but rather copies/structurally shares in the same way as IPLD.

Reply:
lazy languages :raised_hands: